### PR TITLE
Add Global deallocation variant (#3516)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #3516

# Rationale for this change
 

This is the first part of supporting zero-copy conversion between `Vec` and `MutableBuffer`. Crucially we need to carry along the `Layout` as it is UB to call realloc or dealloc with a different layout from the one provided to the initial call to `alloc`.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-buffer/src/alloc/mod.rs [hunk 1]
```diff
@@ -135,20 +135,25 @@ pub(crate) enum Deallocation {
     /// An allocation of the given capacity that needs to be deallocated using arrows's cache aligned allocator.
     /// See [allocate_aligned] and [free_aligned].
     Arrow(usize),
```
Comment (tustvold) [inline arrow-buffer/src/alloc/mod.rs:137]: Eventually I could see us removing this in favor of Global, but I kept it for now

File: arrow-buffer/src/alloc/mod.rs [hunk 2]
```diff
@@ -135,20 +135,25 @@ pub(crate) enum Deallocation {
...
-    /// An allocation from an external source like the FFI interface or a Rust Vec.
-    /// Deallocation will happen
+    /// An allocation using the system's default global allocator and the provided layout
+    Global(Layout),
```
Comment (tustvold) [inline arrow-buffer/src/alloc/mod.rs:139]: This does increase the size of `Dellocation` by 8 bytes. In practice this is extremely unlikely to matter

**Review by viirya — APPROVED** [review]

**Comment by tustvold** [comment]
I need to think a bit more on this, the challenge is Vec will call dealloc with the alignment it would ask for, not the alignment that is actually correct :thinking:

**Comment by tustvold** [comment]
I plan to handle this differently

**Pull request closed without merging**